File: java/PlusOne.java
/*
 * Filename: PlusOne.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 19 May 2015 01:52:02 PM EDT
 */

public class Solution {
	public int[] plusOne(int[] digits) {
		if (digits == null || digits.length == 0) {
			return new int[]{ 1 };
		}

		if (isAll9(digits)) {
			int[] res = new int[digits.length + 1];
			for (int i = 0; i < res.length; i++) {
				if (i == 0) {
					res[i] = 1;
				} else {
					res[i] = 0;
				}
			}
			return res;
		}

		int ditit;
		int carry = 1;
		for (int i = digits.length - 1; i >= 0; i--) {
			digit = carry + digits[i];
			carry = digit / 10;
			digits[i] = digit % 10;
		}

		return digits;
	}

	private boolean isAll9(int[] digits) {
		for (int i = 0; i < digits.length; i++) {
			if (digits[i] != 9) {
				return false;
			}
		}

		return true;
	}
}

File: cpp/plusOne.cpp
/*
 * Filename: plusOne.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Wed 27 Jan 2016 08:35:13 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution
{
	public:
		vector<int> plusOne(vector<int>& digits)
		{
			int carry = 1;

			for (int i = digits.size() - 1; i >= 0; i--)
			{
				carry += digits[i];
				digits[i] = carry % 10;
				carry /= 10;
			}

			if (carry)
			{
				digits.insert(digits.begin(), carry);
			}

			return digits;
		}
};
